File: rxjava/src/main/java/com/asemlab/samples/rxjava/ui/home/CountriesAdapter.java
package com.asemlab.samples.rxjava.ui.home;

import android.os.Build;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.asemlab.samples.rxjava.databinding.ItemCountryBinding;
import com.asemlab.samples.rxjava.models.CountryResponseItem;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class CountriesAdapter extends RecyclerView.Adapter<CountriesAdapter.CountriesVH> {

    private final ArrayList<CountryResponseItem> data;
    private final Consumer<CountryResponseItem> clickListener;

    public CountriesAdapter(List<CountryResponseItem> items, Consumer<CountryResponseItem> listener) {
        data = new ArrayList<>(items);
        clickListener = listener;
    }

    public void updateItems(List<CountryResponseItem> items) {
        data.clear();
        data.addAll(items);
        notifyDataSetChanged();
    }


    @NonNull
    @Override
    public CountriesVH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return CountriesVH.inflate(LayoutInflater.from(parent.getContext()), parent);
    }

    @Override
    public void onBindViewHolder(@NonNull CountriesVH holder, int position) {
        holder.bind(data.get(position), clickListener);
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    static class CountriesVH extends RecyclerView.ViewHolder {

        private ItemCountryBinding itemView;

        private CountriesVH(ItemCountryBinding itemView) {
            super(itemView.getRoot());
            this.itemView = itemView;
        }

        static CountriesVH inflate(LayoutInflater inflater, ViewGroup parent) {
            return new CountriesVH(ItemCountryBinding.inflate(inflater, parent, false));
        }

        public void bind(CountryResponseItem item, Consumer<CountryResponseItem> clickListener) {
            itemView.setCountry(item);
            itemView.getRoot().setOnClickListener(v -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    clickListener.accept(item);
                }
            });
        }
    }

}

File: graphql/src/main/java/com/asemlab/samples/graphql/ui/CountriesAdapter.kt
package com.asemlab.samples.graphql.ui

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.asemlab.samples.graphql.databinding.CountryItemBinding
import com.asemlab.samples.graphql.model.SimpleCountry

class CountriesAdapter(
    private val items: MutableList<SimpleCountry>,
    private val onClick: (SimpleCountry) -> Unit
) : RecyclerView.Adapter<CountriesAdapter.CountriesVH>() {


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CountriesVH {
        return CountriesVH.from(parent)
    }

    override fun onBindViewHolder(holder: CountriesVH, position: Int) {
        holder.bind(items[position], onClick)
    }

    override fun getItemCount() = items.size

    fun updateData(newData: List<SimpleCountry>) {
        items.clear()
        items.addAll(newData)
        notifyDataSetChanged()
    }

    class CountriesVH private constructor(private val binding: CountryItemBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bind(country: SimpleCountry, onClick: (SimpleCountry) -> Unit) {
            with(binding) {
                this.country = country
                root.setOnClickListener {
                    onClick(country)
                }
            }
        }

        companion object {
            fun from(parent: ViewGroup): CountriesVH {
                val binding =
                    CountryItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
                return CountriesVH(binding)
            }
        }

    }

}